Addition of Go FinishWithFileIdentifier

husobee opened the pull request:
Needed the ability to include a file identifier in the flatbuffer output.  These changes will allow the Go implementation to perform that, without changing current functionality.

husobee commented:
I signed it!

aardappel commented on go/builder.go:
```diff
@@ -540,6 +543,24 @@ func (b *Builder) Slot(slotnum int) {
+// FinishWithFileIdentifier finalizes a buffer, pointing to the given `rootTable`.
+// as well as applys a file identifier
+func (b *Builder) FinishWithFileIdentifier(rootTable UOffsetT, fid []byte) {
+	if fid != nil && len(fid) == fileIdentifierLength {
```
rather than silently failing if the length is incorrect, this should probably be a panic?

aardappel commented on go/builder.go:
```diff
@@ -540,6 +543,24 @@ func (b *Builder) Slot(slotnum int) {
+		// to prepare an alignment and file identifier length
+		b.Prep(b.minalign, SizeInt32+fileIdentifierLength)
+		// zero out for alignment
+		b.PlaceInt32(0)
```
why are these 0-bytes written? these are not written in the other implementations

husobee commented:
@aardappel: updated as requested, added a panic on the failed file identifier size validation.  Also updated call to prep to perform padding which aligns with the js and java implementations.

aardappel commented on go/builder.go:
```diff
@@ -540,6 +545,24 @@ func (b *Builder) Slot(slotnum int) {
+	}
+	// In order to add a file identifier to the flatbuffer message, we need
+	// to prepare an alignment and file identifier length
+	b.Prep(SizeInt32+fileIdentifierLength, 0)
```
This now aligns to a multiple of 8, which is not necessary, as this needs to align to 4. So just pass either one.

husobee replied:
It seems that the other implementations do this as well.  For example: 
https://github.com/google/flatbuffers/blob/master/java/com/google/flatbuffers/FlatBufferBuilder.java#L858
https://github.com/google/flatbuffers/blob/master/js/flatbuffers.js#L670-L671

Also, flatcc generated verification fails without the alignment a multiple of 8.

aardappel replied:
No, that code is not quite the same. In Java, it says, "after I write these data items, I need the alignment to be `minalign`" (which is 4, or 8 if there's any long/double fields). Your code says "after I write 0 items, align to 8" (which is usually correct, but wasteful). So it be better to copy what Java does, and use minalign.

aardappel commented on go/builder.go:
```diff
@@ -540,6 +545,24 @@ func (b *Builder) Slot(slotnum int) {
+// as well as applys a file identifier
+func (b *Builder) FinishWithFileIdentifier(rootTable UOffsetT, fid []byte) {
+	if fid == nil || len(fid) != fileIdentifierLength {
+		panic(fmt.Sprintf(
```
If you made this "incorrect file identifier length" you can drop the `fmt` dependency (and who knows how much code).

husobee replied:
updated

aardappel commented:
Ok, looking good, thanks!
